### PR TITLE
Replace bundled JSON parser with json-c

### PR DESCRIPTION
**Pull request by ernestask**
This commit drops the bundled JSON parser from
https://github.com/udp/json-parser/ and replaces it with json-c, an
external library. It’s already used in libreport, so we’re not breaking
new ground here by adding alien technology.

With this, we don’t have to care about testing the actual parser,
dealing with any security issues that may arise or otherwise maintaining
the code. All that and less code to compile.

Macros were removed where it didn’t add too much code, because otherwise
debugging is a massive pain in the hindside. The craziest ones remain.

Signed-off-by: Ernestas Kulik <ekulik@redhat.com>

### CODE REVIEW

**Comment by ernestask** [comment]
Leaving as a draft so I could give it one last look-over.

File: lib/json_utils.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    json_read_uint32(object, key_name, dest, error_message)
+#define JSON_READ_UINT16(object, key_name, dest) \
+    json_read_uint16(object, key_name, dest, error_message)
+#define JSON_READ_STRING(object, key_name, dest) \
+    json_read_string(object, key_name, dest, error_message)
+#define JSON_READ_BOOL(object, key_name, dest) \
+    json_read_bool(object, key_name, dest, error_message)
+
+struct sr_strbuf *
+sr_json_append_escaped(struct sr_strbuf *strbuf, const char *str);
```
Comment (mgrabovsky) [inline lib/json_utils.h]: Why not have this header in `include/`?

File: lib/json_utils.h [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <json.h>
+#include <stdbool.h>
+
+/* Checks whether json value is of the right type and sets the error_message if
+ * not. The "name" argument is used in the error message. */
```
Comment (mgrabovsky) [inline lib/json_utils.h]: Could you put the comment terminator on a separate line?

File: lib/json_utils.c [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+    if (json_object_is_type (object, type))
+    {
+        return true;
+    }
+
+    if (error_message)
+    {
+        const char *type_name;
+
+        type_name = json_type_to_name (type);
```
Comment (mgrabovsky) [inline lib/json_utils.c]: Extra space after function name. Same on line 51 above.

Also, wouldn't it be better to squash the declaration and assignment into one statement?

File: lib/java_thread.c [hunk 4]
```diff
@@ -417,22 +417,28 @@ sr_java_thread_from_json(struct sr_json_value *root, char **error_message)
...
+            json_object *frame_json;
+            struct sr_java_frame *frame;
 
+            frame_json = json_object_array_get_idx(frames, i);
+            frame = sr_java_frame_from_json(frame_json, error_message);
             if (!frame)
                 goto fail;
 
-            result->frames = sr_java_frame_append(result->frames, frame);
+            result->frames = sr_java_frame_append(result->frames, frame);
```
Comment (mgrabovsky) [inline lib/java_thread.c]: Trailing whitespace.

File: lib/generic_stacktrace.c [hunk 5]
```diff
@@ -102,23 +102,35 @@ sr_stacktrace_parse(enum sr_report_type type, const char *input, char **error_me
...
+    enum json_tokener_error error;
+    json_object *json_root = json_tokener_parse_verbose(input, &error);
 
     if (!json_root)
+    {
+        if (NULL != error_message)
+        {
+            const char *description;
+
+            description = json_tokener_error_desc(error);
```
Comment (mgrabovsky) [inline lib/generic_stacktrace.c]: Why not squash the declaration and assignment?

File: lib/koops_stacktrace.c [hunk 6]
```diff
@@ -634,22 +634,34 @@ sr_koops_stacktrace_from_json(struct sr_json_value *root, char **error_message)
...
                 goto fail;
 
+            flag = json_object_get_string(flag_json);
+
             for (struct sr_taint_flag *f = sr_flags; f->name; f++)
             {
-                if (0 == strcmp(f->name, flag_json->u.string.ptr))
+                if (0 == strcmp(f->name, flag))
                 {
+                    /* WTF? */
```
Comment (mgrabovsky) [inline lib/koops_stacktrace.c]: 👀

File: lib/report.c [hunk 7]
```diff
@@ -407,28 +412,32 @@ sr_report_from_json(struct sr_json_value *root, char **error_message)
...
 
-        const unsigned children = json_object_children_count(extra);
+        if (!json_check_type(extra, json_type_object, "auth", error_message))
+            goto fail;
 
         /* from the last children down to the first for easier testing :)
          * keep it as it is as long as sr_report_add_auth() does LIFO */
-        for (unsigned i = 1; i <= children; ++i)
+        object = json_object_get_object(extra);
+        for (struct lh_entry *entry = object->tail; NULL != entry; entry = entry->prev)
```
Comment (mgrabovsky) [inline lib/report.c]: There's an [object iterator interface](https://json-c.github.io/json-c/json-c-0.13.1/doc/html/json__object__iterator_8h.html) which seems like a better approach than meddling with the internals.

File: lib/generic_stacktrace.c (hunk 5 above)
Comment (ernestask) [reply lib/generic_stacktrace.c]: That’s just poor form, IMO, mixing code and declarations.

File: lib/json_utils.h (hunk 2 above)
Comment (ernestask) [reply lib/json_utils.h]: The code was copied, didn’t really care about whatever demons were holed up there.

Will do.

File: lib/json_utils.h (hunk 1 above)
Comment (ernestask) [reply lib/json_utils.h]: These are internal utilities, in short.

File: lib/report.c (hunk 7 above)
Comment (ernestask) [reply lib/report.c]: Can it go in reverse?

File: lib/java_thread.c (hunk 4 above)
Comment (ernestask) [reply lib/java_thread.c]: Oops.

File: lib/json_utils.c (hunk 3 above)
Comment (ernestask) [reply lib/json_utils.c]: Muscle memory. Left a comment about structuring code elsewhere.

File: lib/generic_stacktrace.c (hunk 5 above)
Comment (mgrabovsky) [reply lib/generic_stacktrace.c]: OK, that is your prerogative.

File: lib/report.c (hunk 7 above)
Comment (mgrabovsky) [reply lib/report.c]: Hm, right, it doesn't seem to. Though it appears that order isn't critical for `auth_entries`, it's not a huge issue either.

**Comment by mgrabovsky** [comment]
Looks good to merge. Can you please rebase onto master?

**Comment by ernestask** [comment]
Will do a bit later.

**Comment by ernestask** [comment]
Rebaserino complete.

**Pull request merged**